src/oic/utils/session_backend.py

```diff
@@ -0,0 +1,190 @@
+
+    @abstractmethod
+    def __setitem__(self, key: str, value: Dict[str, str]) -> None:
+        """Store the session information under the session_id."""
+
+    @abstractmethod
+    def __getitem__(self, key: str) -> Dict[str, str]:
+        """
+        Retrieve the session information based os session_id.
+
+        @raises KeyError when no key is found.
+        """
+
+    @abstractmethod
+    def __delitem__(self, key: str) -> None:
+        """Remove the stored session from storage."""
+
+    @abstractmethod
+    def __contains__(self, key: str) -> bool:
+        """Test presence of key in storage."""
+
+    @abstractmethod
+    def get_by_uid(self, uid: str) -> List[str]:
+        """Return session ids (keys) based on `uid` (internal user identifier)."""
+
+    @abstractmethod
+    def get_by_sub(self, sub: str) -> List[str]:
+        """Return session ids based on `sub` (external user identifier)."""
+
+    @abstractmethod
+    def get(self, attr: str, val: str) -> List[str]:
+        """Return session ids based on attribute name and value."""
+
+    def get_client_ids_for_uid(self, uid: str) -> List[str]:
+        """Return client ids that have a session for given uid."""
+        return [self[sid]["client_id"] for sid in self.get_by_uid(uid)]
+
+    def get_verified_logout(self, uid: str) -> Optional[str]:
+        """Return logout verification key for given uid."""
+        # Since all the sessions should be the same, we return the first one
+        sids = self.get_by_uid(uid)
+        if len(sids) == 0:
+            return None
+        _dict = self[sids[0]]
+        if "verified_logout" not in _dict:
+            return None
+        return _dict["verified_logout"]
+
+    def get_token_ids(self, uid: str) -> List[str]:
+        """Return id_tokens for the given uid."""
+        return [self[sid]["id_token"] for sid in self.get_by_uid(uid)]
+
+    def is_revoke_uid(self, uid: str) -> bool:
+        """Return if the session is revoked."""
+        # We do not care which session it is - once revoked, al are revoked
+        return any([self[sid]["revoked"] for sid in self.get_by_uid(uid)])
+
+    def update(self, key: str, attribute: str, value: Any):
+        """
+        Update information stored. If the key is not know a new entry will be constructed.
+
+        :param key: Key to the database
+        :param attribute: Attribute name
+        :param value: Attribute value
+        """
+        if key not in self:
+            self[key] = {attribute: value}
+        else:
+            item = self[key]
+            item[attribute] = value
+            self[key] = item
+
+    def get_uid_by_sub(self, sub):
+        """Return User ids based on sub. Should only be one."""
+        uid = ""
+        for sid in self.get_by_sub(sub):
+            if uid:
+                if uid != AuthnEvent.from_json(self[sid]["authn_event"]).uid:
+                    raise ValueError("More the one uid bound to one sub")
```

Comment:
This raise is not tested. Can this really happen in real life?

Reply:
Unlikely, so we could leave it out until we've seen it happen ?

Reply:
I guess... I am for removing this piece of code.

Reply:
OK